src/adapters/EyevinnAdapter.ts

```diff
@@ -0,0 +1,197 @@
+import { Adapter, AdapterConnectOptions } from './Adapter'
+
+const DEFAULT_CONNECT_TIMEOUT = 2000;
+
+interface EyevinnCandidateRequest {
+  candidate: string;
+}
+
+interface EyevinnAnswerRequest {
+  answer: string;
+}
+
+interface EyevinnMediaStream {
+  streamId: string;
+}
+
+interface EyevinnOfferResponse {
+  offer: string;
+  mediaStreams: EyevinnMediaStream[];
+}
+
+export class EyevinnAdapter implements Adapter {
+  private localPeer: RTCPeerConnection;
+  private channelUrl: URL;
+  private debug: boolean;
+  private iceGatheringTimeout: any;
+  private waitingForCandidates: boolean = false;
+  private resourceUrl: URL | undefined = undefined;
+
+  constructor(peer: RTCPeerConnection, channelUrl: URL) {
+    this.localPeer = peer;
+    this.channelUrl = channelUrl;
+    this.debug = true;
+
+    this.localPeer.onicegatheringstatechange = this.onIceGatheringStateChange.bind(this);
+    this.localPeer.oniceconnectionstatechange =
+      this.onIceConnectionStateChange.bind(this);
+    this.localPeer.onicecandidateerror = this.onIceCandidateError.bind(this);
+    this.localPeer.onicecandidate = this.onIceCandidate.bind(this);
+  }
+
+  setupDataChannels(labels: string[]): RTCDataChannel[] {
+    const channels = labels.map(label => this.localPeer.createDataChannel(label));
+    return channels;
+  }
+
+  enableDebug() {
+    this.debug = true;
+  }
+
+  getPeer(): RTCPeerConnection {
+    return this.localPeer;
+  }
+
+  async connect(opts?: AdapterConnectOptions) {
```

Comment:
Moved all SDP exchange handling to specialization and made Adapter a pure interface since the offer/answer model might not be the same for all adapters.